File: Facebook/90_subsetII.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        ##### different from 78, need to sort the nums first ######
        nums.sort()
        self.helper(nums, 0, [], res)
        return res
    
    def helper(self, nums, index, path, res):
        res.append(path)
        for i in range(index, len(nums)):
            ##### different from 78, need to avoid duplicated value ######
            if i != index and nums[i] == nums[i-1]: continue
            self.helper(nums, i+1, path+[nums[i]], res)

File: Leetcode_SecondPass/Backtracking/90_SubsetII.java
class Solution {
    // O(2 ^ n), O(n)     
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0) return res;
        Arrays.sort(nums);
        helper(res, new ArrayList<Integer>(), nums, 0);
        return res;        
    }
    public static void helper(List<List<Integer>> res, List<Integer> current, int[] nums, int index){
        res.add(new ArrayList<>(current));
        for(int i = index; i < nums.length; i++){
            if (i != index && nums[i] == nums[i-1]) continue;
            current.add(nums[i]);
            helper(res, current, nums, i + 1);
            current.remove(current.size()-1);
        }
    }
}